lab-3

PtichnikovEA opened the pull request:
Вопрос: Исходя из кода подумайте когда методы __str__ и __repr__ могут быть унаследованы, а когда перегружены в дочерних классах. И исправьте это:

Ответ: Методы __str__ и __repr__ могут быть унаследованы в тех случаях, когда возвращаемое ими значение будет удовлетворять необходимому для дочернего класса. В данном случае в дочерних классах добавляется новый аттрибут при инициализации, поэтому эти методы должны быть перегружены.

PtichnikovEA committed:
lab-3

Вопрос: Исходя из кода подумайте когда методы __str__ и __repr__ могут быть унаследованы, а когда перегружены в дочерних классах. И исправьте это:

Ответ: Методы __str__ и __repr__ могут быть унаследованы в тех случаях, когда возвращаемое ими значение будет удовлетворять необходимому для дочернего класса. В данном случае в дочерних классах добавляется новый аттрибут при инициализации, поэтому эти методы должны быть перегружены.

frek17 commented on Лабораторная работа 3/task1/task1.py:
```diff
@@ -0,0 +1,65 @@
+        self.aswe = 1
+
+    def __str__(self):
+        return f"Книга {self.name}. Автор {self.author}"
```
Воот, `__str__` таким и нужно оставить в дочерних классах

frek17 commented on Лабораторная работа 3/task1/task1.py:
```diff
@@ -0,0 +1,65 @@
+    def __init__(self, name: str, author: str):
+        self._name = name
+        self._author = author
+        self.aswe = 1
```
Что такое `aswe`?

PtichnikovEA replied:
> Что такое `aswe`?

Опечатка, этого не должно было быть.

frek17 commented on Лабораторная работа 3/task1/task1.py:
```diff
@@ -0,0 +1,65 @@
+        super().__init__(name, author)
+        if type(pages) == int:
+            if pages >= 0:
+                self.pages = pages
```
Это должно быть реализовано в виде свойства, проверка должна быть там же

frek17 commented on Лабораторная работа 3/task1/task1.py:
```diff
@@ -0,0 +1,65 @@
+        super().__init__(name, author)
+        if type(duration) == float or int:
+            if duration >= 0:
+                self.duration = duration
```
Аналогично

frek17 commented on Лабораторная работа 3/task1/task1.py:
```diff
@@ -0,0 +1,65 @@
+if __name__ == '__main__':
+    b = PaperBook('Python for professionals', 'Matt Telles', 23)
+    b.author = 'Harry Rogers'
+    # AttributeError: can't set attribute
```
О чем говорит комментарий?

PtichnikovEA replied:
При попытке изменить автора будет получена ошибка.

frek17 commented:
Пока что 6 из 10

frek17 commented:
И да, `pages`, `duration` должны быть с сеттером

PtichnikovEA committed:
Исправления лабораторной работы 3

PtichnikovEA committed:
Исправление

frek17 commented on Лабораторная работа 3/task1/task1.py:
```diff
@@ -0,0 +1,83 @@
+
+    @name.setter
+    def name(self, value):
+        raise AttributeError('Attribute name is read-only')
```
Setter не нужен

frek17 commented on Лабораторная работа 3/task1/task1.py:
```diff
@@ -0,0 +1,83 @@
+
+    @author.setter
+    def author(self, value):
+        raise AttributeError('Attribute author is read-only')
```
Аналогично

frek17 commented on Лабораторная работа 3/task1/task1.py:
```diff
@@ -0,0 +1,83 @@
+    def pages(self, value):
+        if type(value) == int:
+            if value >= 0:
+                self.pages = value
```
Вы запускали код? Эта строчка правильная?

frek17 commented on Лабораторная работа 3/task1/task1.py:
```diff
@@ -0,0 +1,83 @@
+class PaperBook(Book):
+    def __init__(self, name: str, author: str, pages: int):
+        super().__init__(name, author)
+        self._pages = pages
```
Лучше через setter сделать записать

frek17 commented on Лабораторная работа 3/task1/task1.py:
```diff
@@ -0,0 +1,83 @@
+class AudioBook(Book):
+    def __init__(self, name: str, author: str, duration: float):
+        super().__init__(name, author)
+        self._duration = duration
```
Аналогично

frek17 commented on Лабораторная работа 3/task1/task1.py:
```diff
@@ -0,0 +1,83 @@
+    def duration(self, value):
+        if type(value) == float or int:
+            if value >= 0:
+                self.duration = value
```
Аналогично

frek17 commented:
Конечно, лучше стало, но всё ещё есть замечания

PtichnikovEA committed:
исправления

1. Удалены сеттеры для свойств name и author
2. Исправлены сеттеры для duration и pages
3. Исправлена функция инициализации для PaperBook и AudioBook

frek17 commented:
Увидел исправления, добавляю баллы
